File: Class22_BinarySearch_Strings/NthRootOfNumberPrecised.cpp
#include<iostream>
#include<vector>
using namespace std;

double multiply(double el, int n){
	double prod = 1.0;
	while(n--){
		prod *= el;
	}
	return prod;
}

double nthRootUsingBinarySearch(int m, int n){

	double beg = 1.0;
	double en = m;
	double delta = 0.00001;	// 10^-1

	// beg <= en -> 0 <= en - beg
	while(en - beg > delta){
		double mid = (beg + en) / 2.0;

		double prod = multiply(mid, n);

		if(prod < m){
			beg = mid;
		}
		else{
			en = mid;
		}
	}

	// cout<<beg<<" "<<en<<endl;
	return en;
}

int main(){

	int m;
	cin>>m;

	int n;
	cin>>n;

	// find nth root of m

	// m = 1024, n = 4 -> 4th root of 1024

	cout<<n<<"th Root of "<<m<<" = "<<nthRootUsingBinarySearch(m, n)<<endl;

	return 0;
}

File: Class22_BinarySearch_Strings/NthRootOfNumberPrecised.java
import java.util.Scanner;

public class NthRootOfNumberPrecised{

	public static double multiply(double a, int b){
        double ans = 1;
        for(int i=0; i<b; i++){
            ans = ans * a;
        }
        
        return ans;
    }

	public static double nthRootOfNumber(double no, int n){
		double low = 1.0;
		double high = no;
		double delta = 0.000001;

		while(high - low > delta){
			double mid = (low + high) / 2.0;
			// System.out.println("Low = " + low + " | High = " + high + " | Mid = " + mid + " | Diff = " + (high - low));

			double mid_power_n = multiply(mid, n);

			if(mid_power_n < no){
				low = mid;
			}
			else{
				high = mid;
			}
		}

		return low;
	}

	public static void main(String[] args){
		Scanner scanner = new Scanner(System.in);
		int n = scanner.nextInt();
		double m = scanner.nextDouble();

		System.out.print(n + "th Root of " + m + " = ");
		System.out.printf("%.6f\n", nthRootOfNumber(m, n));
	}
}

